pkg/lightning/backend/local/local.go

```diff
@@ -258,25 +316,330 @@ func (e *File) unlock() {
+						finSeq = metas.seq
+						if len(inSyncSeqs.arr) > 0 {
+							for inSyncSeqs.arr[0] == finSeq+1 {
+								finSeq++
+								heap.Remove(inSyncSeqs, 0)
+								if len(inSyncSeqs.arr) == 0 {
+									break
+								}
+							}
+						}
+						var flushChans []chan struct{}
+						for _, seq := range flushQueue {
+							if seq.seq <= finSeq {
+								flushChans = append(flushChans, seq.ch)
+							} else {
+								break
+							}
+						}
+						flushQueue = flushQueue[len(flushChans):]
+						finishedSeq.Store(finSeq)
+						seqLock.Unlock()
+						if len(flushChans) > 0 {
+							for _, c := range flushChans {
+								c <- struct{}{}
+							}
+						}
+					} else {
+						heap.Push(inSyncSeqs, metas.seq)
+						seqLock.Unlock()
+					}
+				}
+			}
+		}()
+	}
+
+	compactAndIngestSSTs := func(metas []*sstMeta) {
+		if len(metas) > 0 {
+			seqLock.Lock()
+			metaSeq := seq.Add(1)
+			seqLock.Unlock()
+			select {
+			case <-e.ctx.Done():
+			case metaChan <- metaAndSeq{metas: metas, seq: metaSeq}:
+			}
+		}
+	}
+
+	pendingMetas := make([]*sstMeta, 0, 16)
+	totalSize := int64(0)
+	addMetas := func(metas []*sstMeta) {
+		if len(metas) == 0 {
+			return
+		}
+		if !e.config.Compact {
+			compactAndIngestSSTs(metas)
+			return
+		}
+		for _, m := range metas {
+			if m.totalCount > 0 {
+				pendingMetas = append(pendingMetas, m)
+				totalSize += m.totalSize
+				if totalSize >= e.config.CompactThreshold {
+					compactMetas := pendingMetas
+					pendingMetas = make([]*sstMeta, 0, len(pendingMetas))
+					totalSize = 0
+					compactAndIngestSSTs(compactMetas)
+				}
+			}
+		}
+	}
+	metasTmp := make([]*sstMeta, 0)
+	for {
+		closed := false
+		select {
+		case <-e.ctx.Done():
+			close(metaChan)
+			return
+		case m, ok := <-e.sstMetasChan:
+			if !ok {
+				closed = true
+				break
+			}
+			msg := m
+		processMsg:
+			if msg.flushCh != nil {
+				if len(pendingMetas) > 0 {
+					seqLock.Lock()
+					metaSeq := seq.Add(1)
+					flushQueue = append(flushQueue, flushSeq{ch: msg.flushCh, seq: metaSeq})
+					seqLock.Unlock()
+					select {
+					case metaChan <- metaAndSeq{metas: pendingMetas, seq: metaSeq}:
+					case <-e.ctx.Done():
+						close(metaChan)
+						return
+					}
+
+					pendingMetas = make([]*sstMeta, 0, len(pendingMetas))
+					totalSize = 0
+				} else {
+					// none remaining metas needed to be ingested
+					seqLock.Lock()
+					curSeq := seq.Load()
+					finSeq := finishedSeq.Load()
+					// if all pending SST files are written, directly do a db.Flush
+					if curSeq == finSeq {
+						seqLock.Unlock()
+						msg.flushCh <- struct{}{}
+					} else {
+						// waiting for pending compaction tasks
+						flushQueue = append(flushQueue, flushSeq{ch: msg.flushCh, seq: curSeq})
+						seqLock.Unlock()
+					}
+				}
+				break
+			}
+			metasTmp = append(metasTmp, msg.meta)
+			// drain all the sst meta from the chan to make sure all the SSTs are processed before handle a flush msg.
+		inner:
```

Comment:
how about refine this two level loop into a state machine 😵

Reply:
Is there a good example to refer to？😅

Reply:
```golang
state := IDLE
for {
	switch state {
	case CLOSED:
		break
	case IDLE:
		// read a msg
		// set state to NEED_FLUSH or CACHED or CLOSED
	case NEED_FLUSH:
		// send pendingMetas or directly flush
		state = IDLE
	case CACHED:
		// read a msg
		// set state to NEED_FLUSH or CACHED or CLOSED
	}
}
```

not sure current loop could be converted easily 😵

Reply:
And I weakly advise to take a look at this, only read from `e.sstMetasChan` in one place @glorv

Reply:
ok. I'll take a look at this.😂